src/dataflow/src/sink/kafka.rs

```diff
@@ -229,27 +305,112 @@ where
                             connector.frontier.less_equal(&time)
                         };
 
-                        if should_emit {
-                            queue.push_back((row, diff));
+                        if !should_emit {
+                            continue;
+                        }
+
+                        queue.push_back((row, time, diff));
+                        if let Some(consistency) = &mut consistency {
+                            // Note that since a single differential message
+                            // turns into |diff| messages we need to increment
+                            // message counts by |diff| instead of just 1 for
+                            // the consistency topic
+                            let insert =
+                                consistency.update_timestamp_count(time, diff.abs() as i64);
+
+                            if insert {
+                                // Send a BEGIN message for a timestamp the first
```

Comment:
How do we know that this is the first time we're seeing a timestamp?

Reply:
we're tracking timestamps in a hashtable storing a map from Timestamp -> count of how many times we saw the timestamp, and `update_timestamp_count` returns true when we insert a key (aka a Timestamp) for the first time

then we remove timestamps once the input frontier has advanced ahead of that timestamp (basically I see the frontier as a lower bound on what timestamps can come in, and so if we have a frontier of say `{7}` then we know that timestamps `6`, `5` ... are never going to be seen again)